server/tso/tso.go

```diff
@@ -193,40 +176,57 @@ func (t *timestampOracle) isInitialized() bool {
 	return t.tsoMux.tso != nil
 }
 
-// ResetUserTimestamp update the TSO in memory with specified TSO.
-func (t *timestampOracle) ResetUserTimestamp(leadership *election.Leadership, tso uint64) error {
+// resetUserTimestamp update the TSO in memory with specified TSO by an atomicly way.
+func (t *timestampOracle) resetUserTimestamp(leadership *election.Leadership, tso uint64, ignoreSmaller bool) error {
+	t.tsoMux.Lock()
+	defer t.tsoMux.Unlock()
 	if !leadership.Check() {
 		tsoCounter.WithLabelValues("err_lease_reset_ts").Inc()
 		return errs.ErrResetUserTimestamp.FastGenByArgs("lease expired")
 	}
 	nextPhysical, nextLogical := tsoutil.ParseTS(tso)
 	nextPhysical = nextPhysical.Add(updateTimestampGuard)
-	prevPhysical, prevLogical := t.getTSO()
+	var err error
 	// do not update if next logical time is less/before than prev
-	if typeutil.SubTimeByWallClock(nextPhysical, prevPhysical) == 0 && int64(nextLogical) <= prevLogical {
+	if typeutil.SubTimeByWallClock(nextPhysical, t.tsoMux.tso.physical) == 0 && int64(nextLogical) <= t.tsoMux.tso.logical {
 		tsoCounter.WithLabelValues("err_reset_small_counter").Inc()
-		return errs.ErrResetUserTimestamp.FastGenByArgs("the specified counter is smaller than now")
+		if !ignoreSmaller {
+			err = errs.ErrResetUserTimestamp.FastGenByArgs("the specified counter is smaller than now")
+		}
 	}
 	// do not update if next physical time is less/before than prev
-	if typeutil.SubTimeByWallClock(nextPhysical, prevPhysical) < 0 {
+	if typeutil.SubTimeByWallClock(nextPhysical, t.tsoMux.tso.physical) < 0 {
 		tsoCounter.WithLabelValues("err_reset_small_ts").Inc()
-		return errs.ErrResetUserTimestamp.FastGenByArgs("the specified ts is smaller than now")
+		if !ignoreSmaller {
+			err = errs.ErrResetUserTimestamp.FastGenByArgs("the specified ts is smaller than now")
```

Comment:
Will the error be overwritten?

Reply:
The conditions that may go wrong are mutually exclusive and do not occur at the same time.

Reply:
Could the maxResetTSGap be changed?

Reply:
Unless the maxResetTSGap is set to a non-postive value, which I think won't happen and could be found in time during the code review.

Reply:
0 will make line 205 become true

Reply:
I mean according to the meaning of the maxRestTSGap, it couldn't be zero, though we can add a check to make sure its value is positive.

Reply:
IMO, it's better to guarantee this situation will not happen by ourselves. We don't check this value for now.

Reply:
I modified the `adjustDuration` method to make sure the duration value is positive. PTAL.